hibernate-core/src/main/java/org/hibernate/dialect/pagination/AbstractLimitHandler.java

```diff
@@ -178,4 +187,31 @@ protected final int getMaxOrLimit(RowSelection selection) {
 			return maxRows;
 		}
 	}
+
+	/**
+	 * The offset/limit clauses typically must come
+	 * before the {@code FOR UPDATE}ish clauses, so
+	 * we need a way to identify these clauses in
+	 * the text of the whole query.
+	 */
+	protected Pattern getForUpdatePattern() {
+		return FOR_UPDATE_PATTERN;
+	}
+
+	/**
+	 * Insert a fragment of SQL right before the
+	 * {@code FOR UPDATE}ish clauses at the end
+	 * of the query.
+	 */
+	protected String insertBeforeForUpdate(String limitOffsetClause, String sqlStatement) {
+		Matcher forUpdateMatcher = getForUpdatePattern().matcher( sqlStatement );
+		if ( forUpdateMatcher.find() ) {
+			return new StringBuilder( sqlStatement )
```

Comment:
by default, StringBuilder will allocate 16 additional chars after the constructor string parameter. It is very likely the 16 chars are enough for 'limitOffsetClaus'. To avoid StringBuilder resizing, the following approach might work:
```
return new StringBuilder( sqlStatement.size() + limitOffsetClause.size() ).append( sqlStatement )...;
```
But it sacrifices code readability a little bit, so it might not be worthwhile though. Not very sure here.